src/java.base/share/classes/jdk/internal/foreign/FunctionDescriptorImpl.java

```diff
@@ -0,0 +1,238 @@
+     *     <li>the two function descriptors have argument layouts that are pair-wise {@linkplain MemoryLayout#equals(Object) equal}; and</li>
+     *     <li>the two function descriptors have the same leading {@linkplain #firstVariadicArgumentIndex() variadic argument index}</li>
+     * </ul>
+     *
+     * @param other the object to be compared for equality with this function descriptor.
+     * @return {@code true} if the specified object is equal to this function descriptor.
+     */
+    @Override
+    public final boolean equals(Object other) {
+        return other instanceof FunctionDescriptorImpl f &&
+                Objects.equals(resLayout, f.resLayout) &&
+                Objects.equals(argLayouts, f.argLayouts) &&
+                firstVariadicArgumentIndex() == f.firstVariadicArgumentIndex();
+    }
+
+    /**
+     * {@return the hash code value for this function descriptor}
+     */
+    @Override
+    public final int hashCode() {
+        return Objects.hash(argLayouts, resLayout, firstVariadicArgumentIndex());
+    }
+
+    final void checkIndex(int index) {
+        if (index < 0 || index > argLayouts.size())
+            throw new IllegalArgumentException("Index out of bounds: " + index);
+    }
+
+    public static FunctionDescriptor of(MemoryLayout resLayout, List<MemoryLayout> argLayouts) {
+        return new FunctionDescriptorImpl(resLayout, argLayouts);
+    }
+
+    public static FunctionDescriptor ofVoid(List<MemoryLayout> argLayouts) {
+        return new FunctionDescriptorImpl(null, argLayouts);
+    }
+
+    static final class VariadicFunctionDescriptor extends FunctionDescriptorImpl {
+
+        private final int firstVariadicIndex;
+
+        /**
+         * Constructor.
+         *
+         * @param descriptor the original functional descriptor
+         * @param argLayouts the memory layouts to apply
+         * @throws NullPointerException if any of the provided parameters or array elements are {@code null}
+         */
+        VariadicFunctionDescriptor(FunctionDescriptorImpl descriptor, MemoryLayout... argLayouts) {
+            super(descriptor.returnLayout().orElse(null),
+                    Stream.concat(descriptor.argumentLayouts().stream(), Arrays.stream(argLayouts).map(Objects::requireNonNull)).toList());
+            this.firstVariadicIndex = descriptor.argumentLayouts().size();
+        }
+
+        @Override
+        public int firstVariadicArgumentIndex() {
+            return firstVariadicIndex;
+        }
+
+        @Override
+        public FunctionDescriptorImpl insertArgumentLayouts(int index, MemoryLayout... addedLayouts) {
+            checkIndex(index);
+            for (MemoryLayout layout : addedLayouts) {
+                requireNonNull(layout);
+            }
+            throw newUnsupportedOperationException();
+        }
+
+        @Override
+        public FunctionDescriptorImpl changeReturnLayout(MemoryLayout newReturn) {
+            requireNonNull(newReturn);
```

Comment:
Do we need to enforce other problems (index, null) before throwing UnsupportedOperationException? Existing code unconditionally throws UnsupportedOperationException.